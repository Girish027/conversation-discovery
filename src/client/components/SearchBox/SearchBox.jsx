import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { SearchIcon, Xmark, TextField } from '@tfs/ui-components';
import IconButton from '../IconButton/index';
import { shouldHandle } from '../../utils/KeyboardUtils';
import { searchBarStyle } from '../../styles/index';

export default class SearchBox extends Component {
  constructor(props) {
    super(props);
    this.props = props;

    this.state = {
      value: '',
    };

    this.onSearch = this.onSearch.bind(this);
    this.onClickClear = this.onClickClear.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
  }

  onSearch(event) {
    this.setState({
      value: event.target.value
    }, () => this.props.onSearch(this.state.value));
  }

  onClickClear() {
    this.setState({
      value: '',
    }, () => this.props.onSearch(''));
  }

  onKeyPress(event) {
    if (shouldHandle(event)) {
      this.onSearch(event);
    }
  }

  render() {
    const {
      placeholder, styleOverride, closeIcon
    } = this.props;

    const { value } = this.state;

    // TODO: Need to add width to main container and correct the clear icon styling
    return (
      <div
        style={{
          ...searchBarStyle,
          ...styleOverride
        }}
      >
        <div style={searchBarStyle.searchIcon}>
          <SearchIcon width='12px' fill='#004c97' />
        </div>
        <div style={{ ...searchBarStyle.textFieldContainerStyle, ...styleOverride.container }}>
          <TextField
            name='search-bar'
            onChange={this.onSearch}
            defaultValue={value}
            placeholder={placeholder}
            onKeyPress={this.onKeyPress}
            styleOverride={{ ...searchBarStyle.textFieldStyle, ...styleOverride.textField }}
          />
        </div>
        {closeIcon && value && (
          <div style={{ ...searchBarStyle.clearIcon, ...styleOverride.clearIcon }}>
            <IconButton
              onClick={this.onClickClear}
              icon={Xmark}
              data-qa={'clear-topic'}
              padding={'12px 10px'}
              title='Clear'
            />
          </div>
        )}
      </div>
    );
  }
}

SearchBox.propTypes = {
  placeholder: PropTypes.string,
  onSearch: PropTypes.func,
  styleOverride: PropTypes.object,
  closeIcon: PropTypes.bool,
};

SearchBox.defaultProps = {
  placeholder: '',
  onSearch: () => {},
  styleOverride: {},
  closeIcon: true
};
