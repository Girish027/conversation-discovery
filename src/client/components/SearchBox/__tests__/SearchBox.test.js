import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import SearchBox from '../SearchBox';

describe('<SearchBox />', () => {
  let wrapper;

  const props = {
    placeholder: 'Find Topics',
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  const getShallowWrapper = (propsObj) => shallow(<SearchBox
    {...propsObj}
  />);

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getShallowWrapper(props);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly', () => {
      wrapper = getShallowWrapper(props);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    beforeEach(() => {
      props.onSearch = jest.fn();
      wrapper = getShallowWrapper(props);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    describe('onSearch', () => {
      test('should change the value of searchString', () => {
        const event = {
          target: {
            value: 'New Search'
          }
        };
        expect(wrapper.state().value).toEqual('');
        wrapper.instance().onSearch(event);
        expect(wrapper.state().value).toEqual('New Search');
        expect(props.onSearch).toHaveBeenCalled();
      });
    });

    describe('onClickClear', () => {
      test('should clear the value of searchString', () => {
        wrapper.instance().onClickClear();
        expect(wrapper.state().value).toEqual('');
        expect(props.onSearch).toHaveBeenCalled();
      });
    });

    describe('onEnterPressed', function () {
      test('should update the search value when enter is pressed', () => {
        const event = {
          target: {
            value: 'New Search'
          },
          keyCode: 13,
        };
        wrapper.instance().onKeyPress(event);
        expect(wrapper.state().value).toEqual('New Search');
        expect(props.onSearch).toHaveBeenCalled();
      });
    });
  });
});
