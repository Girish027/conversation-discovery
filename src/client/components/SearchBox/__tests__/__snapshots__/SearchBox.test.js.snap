// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> Snapshots renders correctly 1`] = `
<div
  style={
    Object {
      "clearIcon": Object {
        "position": "relative",
        "right": 30,
      },
      "display": "flex",
      "paddingLeft": "31px",
      "paddingTop": "10px",
      "searchIcon": Object {
        "padding": "12px 10px",
        "position": "absolute",
        "zIndex": 1,
      },
      "textFieldContainerStyle": Object {
        "width": "90%",
      },
      "textFieldStyle": Object {
        "input": Object {
          "paddingLeft": "30px",
        },
        "width": "13vw",
      },
    }
  }
>
  <div
    style={
      Object {
        "padding": "12px 10px",
        "position": "absolute",
        "zIndex": 1,
      }
    }
  >
    <SearchIcon
      fill="#004c97"
      height="1em"
      width="12px"
    />
  </div>
  <div
    style={
      Object {
        "width": "90%",
      }
    }
  >
    <withTheme(TextField)
      defaultValue=""
      name="search-bar"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Find Topics"
      styleOverride={
        Object {
          "input": Object {
            "paddingLeft": "30px",
          },
          "width": "13vw",
        }
      }
    />
  </div>
</div>
`;
